jiant/tasks/lm.py

```diff
@@ -172,3 +179,134 @@ def __init__(self, path, *args, **kw):
         }
+
+
+@register_task("mlm", rel_path="WikiText103/")
+class MaskedLanguageModelingTask(Task):
+    """
+    Masked language modeling task on Wikipedia dataset
+    Attributes:
+        max_seq_len: (int) maximum sequence length
+        min_seq_len: (int) minimum sequence length
+        files_by_split: (dict) files for three data split (train, val, test)
+    """
+
+    def __init__(self, path, max_seq_len, name, **kw):
+        """Init class
+        Args:
+            path: (str) path that the data files are stored
+            max_seq_len: (int) maximum length of one sequence
+            name: (str) task name
+        """
+        super().__init__(name, **kw)
+        self.scorer1 = Average()
+        self.scorer2 = None
+        self._label_namespace = self.name + "_labels"
+        self.val_metric = "%s_perplexity" % self.name
+        self.val_metric_decreases = True
+        self.max_seq_len = max_seq_len
+        self.min_seq_len = 0
+        self.target_indexer = {"words": SingleIdTokenIndexer(namespace="tokens")}
+        self.files_by_split = {
+            "train": os.path.join(path, "train.txt"),
+            "val": os.path.join(path, "valid.txt"),
+            "test": os.path.join(path, "test.txt"),
+        }
+
+    def load_data(self):
+        # Data is exposed as iterable: no preloading
+        self.examples_by_split = {}
+        for split in self.files_by_split:
+            self.examples_by_split[split] = list(self.get_data_iter(self.files_by_split[split]))
+
+    def get_metrics(self, reset=False):
+        """Get metrics specific to the task
+        Args:
+            reset: (boolean) reset any accumulators or internal state
+        """
+        nll = self.scorer1.get_metric(reset)
+        return {"perplexity": math.exp(nll)}
+
+    def get_all_labels(self):
+        """
+        For MLM, the label space is the vocabulary space of the input.
+        """
+        labels = []
+        tokenizer = get_tokenizer(self._tokenizer_name)
+        vocab_size = len(tokenizer)
+        ordered_vocab = tokenizer.convert_ids_to_tokens(range(vocab_size))
+        for word in ordered_vocab:
+            labels.append(word)
+        return labels
+
+    def update_metrics(self, out, batch=None):
+        # self.scorer1(logits,labels)
+        self.scorer1(out["loss"].mean())
+        return
+
+    def get_data_iter(self, path):
+        """Loading data file and tokenizing the text
+        Args:
+            path: (str) data file path
+        """
+        import csv
+
+        f = open(path, "r", encoding="utf-8")
+        reader = csv.reader(f)
+        text = list(reader)
+        moses_tokenizer = get_tokenizer("MosesTokenizer")
+        for i in range(10):
+            row = text[i]
+            untokenized_toks = moses_tokenizer.detokenize(row)
+            toks = "".join(untokenized_toks)
+            yield tokenize_and_truncate(self._tokenizer_name, toks, self.max_seq_len)
```

Comment:
Please refactor to move any necessary imports to the top of the file, ensure that the file object is closed, and avoid converting the file reading generator to a list inside the generator you're building. You can find an example implementation of `get_data_iter()` in `AutoregressiveLanguageModelingTask`.